pricegraph/src/orderbook/iter.rs

```diff
@@ -0,0 +1,203 @@
+    graph::path::Path,
+    orderbook::{self, Flow, Orderbook, OverlapError, ReducedOrderbook},
+};
+use petgraph::graph::NodeIndex;
+use std::{iter::FusedIterator, mem};
+
+/// An iterator over all transitive orders over a token pair for an orderbook,
+/// ordered from lowest limit price to highest.
+pub struct TransitiveOrders {
+    orderbook: Orderbook,
+    pair: Option<(NodeIndex, NodeIndex)>,
+    state: State,
+}
+
+/// Internal iterator state.
+enum State {
+    /// The next transitive order has already been pre-computed and is ready to
+    /// be returned.
+    NextOrder(Option<(Path<NodeIndex>, Flow)>),
+    /// The previous order needs to be filled before continuing.
+    FillPreviousOrder(Path<NodeIndex>, Flow),
+    /// The next transitive order needs to be found, and no previous orders
+    /// need to be filled.
+    Search,
+}
+
+impl TransitiveOrders {
+    /// Creates an empty transitive orderbook iterator.
+    fn empty(orderbook: Orderbook) -> Self {
+        Self {
+            orderbook,
+            pair: None,
+            state: State::Search,
+        }
+    }
+
+    /// Creates a new transitive orderbook iterator.
+    pub fn new(orderbook: Orderbook, pair: TokenPair) -> Result<Self, OverlapError> {
+        let (buy, sell) = if orderbook.is_token_pair_valid(pair) {
+            (
+                orderbook::node_index(pair.buy),
+                orderbook::node_index(pair.sell),
+            )
+        } else {
+            return Ok(Self::empty(orderbook));
+        };
+
+        // NOTE: We need to check that the orderbook is not overlapping in the
+        // subgraph containing the token pair we care about, so we find the
+        // first transitive order and reuse the result in the first call to
+        // `next`.
+        let next_order = orderbook.find_path_and_flow(buy, sell)?;
+
+        Ok(Self {
+            orderbook,
+            pair: Some((buy, sell)),
+            state: State::NextOrder(next_order),
+        })
+    }
+
+    /// Creates a new transitive orderbook iterator from a reduced orderbook.
+    pub fn from_reduced(orderbook: ReducedOrderbook, pair: TokenPair) -> Self {
+        let orderbook = orderbook.into_inner();
+        let pair = if orderbook.is_token_pair_valid(pair) {
+            (
+                orderbook::node_index(pair.buy),
+                orderbook::node_index(pair.sell),
+            )
+        } else {
+            return Self::empty(orderbook);
+        };
+
+        Self {
+            orderbook,
+            pair: Some(pair),
+            state: State::Search,
+        }
+    }
+}
+
+impl Iterator for TransitiveOrders {
+    type Item = Flow;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (buy, sell) = self.pair?;
+        let (path, flow) = match mem::replace(&mut self.state, State::Search) {
+            State::NextOrder(order) => order?,
+            State::FillPreviousOrder(path, flow) => {
+                self.orderbook
```

Comment:
Why do you only want to fill the flow in the next path? It seems a bit brittle if something else was mutating the underlying orderbook in the meantime (I guess this is not likely because of visibility).

Reply:
> Why do you only want to fill the flow in the next path?

This is so we can imitate the behaviour of `fill_transitive_order_if` where we search for it, check some predicate, and then fill it. This allows us two things:
- We get a "look-ahead(1)" and can `break` before removing an order
- We avoid updating the projection graph (a fairly expensive operation) on the last iteration after we break.

> It seems a bit brittle if something else was mutating the underlying orderbook in the meantime

I'm taking full advantage of single ownership and single mutable reference :smile:. Also, the orderbook it taken by value so it can't be mutated by outside.

Reply:
Even with single ownership and single mutable reference, wouldn't I still be able to do something like:

```rust
let mut t = orderbook.clone().transitive_orders(pair).unwrap();
let foo = t.next();
t.orderbook.fill_market_ring_trade(Market {
  base: 4.into(),
  quote: 5.into(),
});
```

Reply:
This would require `orderbook` to be public. What I meant was it would be impossible to have an outside reference to the inner orderbook simultaneously, but yes, it it were exposed as a public property then this would be an issue.